src/util/codec/mysql/json.rs

```diff
@@ -31,66 +35,144 @@ const TYPE_CODE_I64: u8 = 0x09;
 const TYPE_CODE_DOUBLE: u8 = 0x0b;
 const TYPE_CODE_STRING: u8 = 0x0c;
 
+const PRECEDENCE_BLOB: i32 = -1;
+const PRECEDENCE_BIT: i32 = -2;
+const PRECEDENCE_OPAQUE: i32 = -3;
+const PRECEDENCE_DATETIME: i32 = -4;
+const PRECEDENCE_TIME: i32 = -5;
+const PRECEDENCE_DATE: i32 = -6;
+const PRECEDENCE_BOOLEAN: i32 = -7;
+const PRECEDENCE_ARRAY: i32 = -8;
+const PRECEDENCE_OBJECT: i32 = -9;
+const PRECEDENCE_STRING: i32 = -10;
+const PRECEDENCE_NUMBER: i32 = -11;
```

Comment:
IMO it's better to define 
```
PRECEDENCE_INTEGER = -11
PRECEDENCE_DOUBLE = -11
```
to map the real types, it's easier to understand since they are listed in real types in MySQL docs and in tidb's code.